src/cassandra/future.rs

```diff
@@ -1,333 +1,254 @@
-        }
-    }
-
-
-
-    /// Gets the result of a successful future. If the future is not ready this method will
-    /// wait for the future to be set.
-    /// a None response indicates that there was an error
-    pub fn get(&mut self) -> Option<CassResult> {
-        unsafe {
-            let result = cass_future_get_result(self.0);
-            if result.is_null() {
-                None
-            } else {
-                Some((CassResult::build(result)))
+        };
+
+        if install_callback {
+            // Install the callback. If callback cannot be sent, report immediate `Err`.
+            let data =
+                (self.state.as_ref() as *const FutureTarget) as *mut ::std::os::raw::c_void;
+            cass_future_set_callback(self.inner, Some(notify_task), data)
+                .to_result(()).chain_err(|| ErrorKind::DriverError("Unable to set callback"))
+        } else {
+            Ok(())
+        }.and_then(move |_| ret)
+    }}
+}
+
+/// Extract success or failure from a completed future.
+unsafe fn get_completion<T: Completable>(inner: *mut _Future) -> Result<T> {
+    // Future is ready; wrap success in `Ok(Ready)` or report failure as `Err`.
+    let rc = cass_future_error_code(inner);
+    match rc {
+        CASS_OK => {
+            match Completable::get(inner) {
+                None => Err(ErrorKind::DriverError("No result found").into()),
+                Some(v) => Ok(v)
             }
-        }
+        },
+        _ => Err(get_cass_error(rc, inner)),
     }
 }
 
-
-/// The future result of an prepared statement.
-/// It can represent a result if the operation completed successfully or an
-/// error if the operation failed. It can be waited on, polled or a callback
-/// can be attached.
-#[derive(Debug)]
-pub struct PreparedFuture(*mut _Future);
-
-impl Drop for PreparedFuture {
-    fn drop(&mut self) { unsafe { cass_future_free(self.0) } }
-}
-
-impl PreparedFuture {
-    /// Wait for the future to be set with either a result or error.
-    ///
-    /// Important: Do not wait in a future callback. Waiting in a future
-    /// callback will cause a deadlock.
-    pub fn wait(&mut self) -> Result<PreparedStatement> {
-        unsafe {
-            cass_future_wait(self.0);
-            self.error_code()
-        }
-    }
-
-    /// Gets the error code from future. If the future is not ready this method will
-    /// wait for the future to be set.
-    pub fn error_code(&mut self) -> Result<PreparedStatement> {
-        unsafe { cass_future_error_code(self.0).to_result(self.get()).chain_err(|| "") }
+/// Extract the error code and message from a Cassandra driver future
+unsafe fn get_cass_error(rc: CassError_, inner: *mut _Future) -> Error {
+    let e = cass_future_get_error_result(inner);
+    if e.is_null() {
+        // No extended error available; just take the basic one.
+        let mut msg = mem::zeroed();
+        let mut msg_length = mem::zeroed();
```

Comment:
The philosophy here of course is that what you don't write, you can't get wrong. If there's no `malloc(4)`, then you can never get into trouble because the `4` should have been `8`.